nixos/modules/virtualisation/containers.nix

```diff
@@ -113,6 +150,8 @@ in
 
   config = lib.mkIf cfg.enable {
 
+    hardware.nvidia-container-toolkit.enable = lib.mkIf cfg.cdi.dynamic.nvidia.enable true;
+
```

Comment:
1. If you provide a conflicting configuraiton, you'll get an error. We can add a warning that is more explicit to the user about what they did wrong. I think `virtualization.containers.cdi.dynamic.<provider>.enable` is the API the user should use. `hardware.nvidia-container-toolkit.enable` is an implementation detail if you wish.
2. The spec details very well what static and dynamic configuration are. I don't see a reason to not obey the spec on the initial implementation just to get into troubling situations in the future when new providers arise and they support automatic CDI generation based on detected hardware. We have the option to design the user-facing API in a way that resembles the feature 1:1, I don't see why we wouldn't. 
    1. With `cdi.${name}` and `hardware.nvidia-container-toolkit.enable` disjoint configurations, you might have conflicting configurations (static vs dynamic at the CDI level, where dynamic have more priority), configuring a single part of the system: it is hard to reason about their relationship from the user perspective. Whereas having `virtualisation.containers.cdi.{static,dynamic}` it's crystal clear what they are about.
    2. This problem will only get worse when more CDI auto-generators based on detected hardware come in.
3. I strongly disagree on that, based on the above.

Reply:
Thanks!

2. If those are concepts from the specification that's a good reason. Following the [link](https://github.com/cncf-tags/container-device-interface/blob/7b540ad63013eafbb6779fd06e3235653528d4ae/pkg/cdi/spec-dirs.go#L26-L31) from your description, I see it does mention there are two directories where configurations are looked for (`/etc/cdi` and `/var/run/cdi`). I see the README also discussing this. It's not clear to me if "CDI" also defines the merge semantics for overlapping static and dynamic configurations or if it's something that your `dynamic.nvidia.enable` takes the responsibility for... but that's irrelevant for the merge I guess


However

1. This does successfully evaluate: https://gist.github.com/SomeoneSerge/d51326df582240b05f4eb59efb7add44. For this _not_ to evaluate you'd have to add a definition setting `...dynamic.enable` whenever the nvdidia ctk service is enable. But you alread have a definition going in the opposite direction, and if `a => b` and `b => a`, `b` and `a` are the same thing so it's redundant to have both.

If we expect to add more services to `cdi.dynamic`, maybe we should rename `cdi.dynamic.nvidia` into e.g. `cdi.dynamic.nvidia-ctk` and establish a naming scheme for the corresponding systemd services (`cdi-dynamic-${name}`, etc)


That said, if we turn things around this way you are right that `hardware.nvidia-container-toolkit` is an implementation detail so this ultimately not a blocker for merging

...deleting it now, we'll save an extra `mkRemovedOptionModule` in the very near future -> save a tiny little bit of eval time

Reply:
The case at https://gist.github.com/SomeoneSerge/d51326df582240b05f4eb59efb7add44 does not concern me too much, given things will "work". The case that in my opinion we want to reject is:

```nix
hardware.nvidia-container-toolkit.enable = false;
virtualisation.containers.cdi.dynamic.nvidia.enable = true;
```

Namely, the "public" interface sets NVIDIA to "magically work", but the user by mistake sets `nvidia-container-toolkit` enable to false, breaking the internals without knowing.

I would not do `cdi.dynamic.nvidia-ctk`. Again, it's an implementation detail. There is no other way to 'enable nvidia automatically' with anything else than 'nvidia-ctk'. The public interface should not have references to anything that resembles internal implementation.

For everything that concerns the user, they know that they do `cdi.dynamic.nvidia.enable = true;` and things happen so that they can use their NVIDIA GPU's on their containers. How that happens is not relevant to them.

Is there some mechanism to mark `hardware.nvidia-container-toolkit.enable` as "private"? So that it can be exclusively set by other modules, and not by the user?

Reply:
> The case that in my opinion we want to reject is:

Why make it representable in the first place... I don't mind if you name the one option `dynamic.nvidia` or `dynamic.nvidia-ctk`, although I do think the latter makes more sense as `nvidia-ctk` is the generator the user wants to use and it will always be easier for us to support concrete options instead of magic/abstract options.

What I do care about is that I still fail to see any reason for `services.nvidia-container-toolkit` to exist. I don't see what it adds to `cdi.dynamic.nvidia${maybe-suffix}`, other than the necessity to implement more asserts and more warnings

Going further with the "abstractions vs narrowly focused modules"... As a user I'd also feel less lost seeing something like `nvidia-ctk-cdi-generate.service` in `systemctl show` than seeing `nvidia-container-toolkit`. Similarly, I think I'd benefit from a generalized naming scheme for the systemd unit that correspond to `cdi.dynamic` entries

> Is there some mechanism to mark hardware.nvidia-container-toolkit.enable as "private"?

I think `internal = true` would attain this end. Although I stated my preferences above.

Reply:
Maybe I'm wrong on all of this... my presumption is that whatever automation and abstraction we implement, it will fail at least on some users, and they would have to override something locally. For that, they'd need to be able to untangle all the complexity and the opaque introduced by abstractions, which is why I'd like to go easy on them

Reply:
I have marked `nvidia-container-toolkit.enable` as internal and not visible (not present in the documentation.) I think this is a good balance and holds both the public facing "API", while keeping separate the implementation detail, but still as a module.

Also, renamed it from `nvidia-container-toolkit` to `nvidia-container-toolkit-cdi-generator`, so if inspected with `systemctl` or `journalctl` it's more descriptive of what it does.

Reply:
Alright, I think with `internal = true` it's OK to keep it for the first iteration. When/if we delete it we won't have to add `mkRenamedOptionModule` because it was internal.

However, as far as I'm concerned, its existence still has not been justified.